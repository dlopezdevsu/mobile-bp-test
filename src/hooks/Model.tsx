class model_class {

  dispatch: any = null
  get_dispatch = () => this.dispatch == null ? false : this.dispatch
  set_dispatch = (n_dispatch: any) => { this.dispatch = n_dispatch }

  setData = (label: string, payload: any) => {
    this?.dispatch({ type: 'SET_STORE', label, payload })
  }

  resetData = () => {
    this?.dispatch({ type: 'RESET_DATA' })
  }
}

export default new model_class();