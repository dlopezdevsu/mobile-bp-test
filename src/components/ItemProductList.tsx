import { useNavigation } from '@react-navigation/native';
import * as React from 'react';
import { View, StyleSheet, Text, Image, Pressable } from 'react-native';
import * as Routes from 'src/utils/routes';

import * as Colors from 'src/utils/colors';
import Model from 'src/hooks/Model';

export interface ItemProductListProps {
  id: string;
  name: string;
  description?: string;
  logo?: string;
  date_release?: string;
  date_revision?: string;
}

const ItemProductList = (product: ItemProductListProps) => {
  const { name, id } = product;
  const navigation = useNavigation()

  const onSelect = () => {
    Model.setData('selected', product)
    navigation.navigate(Routes.PRODUCT_DETAIL as never)
  }

  return (
    <Pressable style={({ pressed }) => [styles.container, { opacity: pressed ? 0.5 : 1 }]} onPress={onSelect}>
      <View style={styles.info}>
        <Text style={styles.name}>{name}</Text>
        <Text style={styles.id}>ID: {id}</Text>
      </View>
      <Image style={styles.icon} source={require('src/assets/icons/chevron-right-icon.png')} />
    </Pressable>
  );
};

const styles = StyleSheet.create({
  container: {
    borderWidth: 1,
    borderColor: Colors.gray,
    backgroundColor: Colors.white,
    padding: 12,
    flexDirection: 'row',
    alignItems: 'center'
  },
  info: {
    flex: 1
  },
  name: {
    fontWeight: 'bold',
  },
  id: {
    marginTop: 6
  },
  icon: {
    resizeMode: 'contain',
    width: 12,
    height: 12,
  },
});

export default ItemProductList;