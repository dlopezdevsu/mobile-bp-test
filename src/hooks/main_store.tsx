import React, { ReactElement } from "react";

import { ItemProductListProps } from 'src/components/ItemProductList'

export const Store = React.createContext<any>(null);

export interface InitialStateProps {
  loading: boolean;
  alert: string;
  products: ItemProductListProps[];
  selected: ItemProductListProps | null;
  editMode: boolean;
  openBottomSheet: boolean;
  callbackBottomSheet: () => void;
}

interface ActionProps {
  label: string;
  type: string;
  payload: any;
}

const initialState = {
  loading: true,
  alert: '',
  products: [],
  selected: null,
  editMode: false,
  openBottomSheet: false,
  callbackBottomSheet: () => null,
};

function reducer(state: InitialStateProps, action: ActionProps) {
  switch (action.type) {
    case "SET_STORE":
      return { ...state, [action.label]: action.payload };
    case "RESET_DATA":
      return { ...initialState };
    default:
      return state;
  }
}

export function StoreProvider({ children }: { children: ReactElement }) {
  const [state, dispatch] = React.useReducer(reducer, initialState);
  const value = { state, dispatch };

  return <Store.Provider value={value}>{children}</Store.Provider>;
}
