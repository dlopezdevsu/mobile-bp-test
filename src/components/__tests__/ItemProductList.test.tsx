/**
 * @format
 */

import 'react-native';
import React from "react";
import ItemProductList from "src/components/ItemProductList";
import renderer from "react-test-renderer";
import { StoreProvider } from 'src/hooks/main_store'

import { it, expect } from '@jest/globals';

const mockedDispatch = jest.fn();
jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: mockedDispatch,
    }),
  };
});

it(`render ItemProductList`, () => {
  const tree = renderer.create(<StoreProvider>
    <ItemProductList id='123' name='Test name' />
  </StoreProvider>);
  expect(tree).toMatchSnapshot();
});
