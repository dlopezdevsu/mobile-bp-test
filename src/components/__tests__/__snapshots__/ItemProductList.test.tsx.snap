// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ItemProductList 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "borderColor": "#eaecf3",
        "borderWidth": 1,
        "flexDirection": "row",
        "padding": 12,
      },
      {
        "opacity": 1,
      },
    ]
  }
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        {
          "fontWeight": "bold",
        }
      }
    >
      Test name
    </Text>
    <Text
      style={
        {
          "marginTop": 6,
        }
      }
    >
      ID: 
      123
    </Text>
  </View>
  <Image
    source={
      {
        "testUri": "../../../src/assets/icons/chevron-right-icon.png",
      }
    }
    style={
      {
        "height": 12,
        "resizeMode": "contain",
        "width": 12,
      }
    }
  />
</View>
`;
